### ISSUE TITLE
[BUG][Fuzzer][ice-on-valid-code] `T.reduce_max`/`reduce_min` into a `float8_e4m3`/`float4` fragment crashes nvcc with undefined `inff` instead of compiling

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2998]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) and [Discussions](https://github.com/tile-ai/tilelang/discussions) that this hasn't already been reported.

### What version of TileLang are you using?

0.1.13 (still present on `main`, inspected, for the `float8_e4m3`/`float4` sub-case — see Additional context).

### System information

TileLang 0.1.13 / CUDA 12.8 / NVIDIA L40S (sm_89).

### Problem description

`T.reduce_max` (or `T.reduce_min`) whose destination fragment is a `float8_e4m3` type fails to compile: nvcc aborts with `identifier "inff" is undefined`. The reduction lowers correctly and torch reduces the same fp8 inputs fine; the failure is purely in the emitted CUDA source.

The reduce lowers an init constant of `-INFINITY` (for max) or `+INFINITY` (for min) into the fp8 accumulator, and codegen prints that infinity as the literal `-inff`:

```
Cf[0] = fp8_e4_t(-inff/*-inf*/);   // nvcc: identifier "inff" is undefined
```

Swapping only the destination fragment to `float32` (same reduce, same fp8 inputs) compiles and returns the correct result, so the crash is specific to printing the fp8-typed infinity init, not to the reduction. `float4_e2m1fn` hits the identical `inff` emission; `float8_e5m2` no longer emits `inff` on 0.1.13 (its inf/nan cases were guarded by #2671) — see Additional context.

Not a regression — the fp8 codegen path has emitted this since it was added (see Provenance).

### Reproducible example code

```python
import tilelang
import tilelang.language as T
import torch

M, N = 32, 64

@tilelang.jit(out_idx=[1])
def kernel(dt):
    @T.prim_func
    def main(A: T.Tensor((M, N), dt), C: T.Tensor((M,), dt)):
        with T.Kernel(1, threads=128):
            Af = T.alloc_fragment((M, N), dt)
            Cf = T.alloc_fragment((M,), dt)
            T.copy(A, Af)
            T.reduce_max(Af, Cf, dim=1)   # init = make_const(dt, -INFINITY)
            T.copy(Cf, C)
    return main

a = torch.rand(M, N, device="cuda")
ref = a.max(1).values

# BUG: fp8 destination -> nvcc "identifier \"inff\" is undefined"
kernel("float8_e4m3")(a.to(torch.float8_e4m3fn))   # -> CUDA compile error

# CONTROL: identical reduce, float32 destination -> compiles, correct
c = kernel("float32")(a)
print((c - ref).abs().max().item())                # -> 0.0
```

### Traceback

```
Compilation error:
tvm_kernels.cu(17): error: identifier "inff" is undefined
    Cf[0] = fp8_e4_t(-inff/*-inf*/);
tvm_kernels.cu(20): error: no operator "=" matches these operands
```

### Expected behavior

The kernel should compile — a max/min reduction over an fp8 fragment is a legal reduction. `float16`/`bfloat16` and `float32`/`float64` destinations already compile the same reduction correctly, because their branches special-case the infinity init (`std::numeric_limits<...>::infinity()` and `CUDART_INF_F`). The fp8/fp4 branch is the one path that emits a raw hexfloat with no inf/nan spelling.

The precedent for the fix direction is in-repo: [#2671](https://github.com/tile-ai/tilelang/pull/2671) added exactly this guard to the fp8 branch for `float8_e5m2` (emitting the inf/nan bit pattern via `::bitcast`). The same guard is still missing for `float8_e4m3`/`float4`, which fall through to the unguarded hexfloat path.

### Additional context

**Root cause.** The CUDA codegen's float-constant printer does not special-case infinity/NaN for the `float8_e4m3`/`float4` sub-branch, so it prints them as an invalid literal. In [`PrintConst`, the fp8/fp4 branch](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/src/cuda/codegen/codegen_cuda.cc#L5776-L5792) the `float8_e5m2` inf/NaN cases are now guarded (`::bitcast(0x7c/0xfc/0x7e)`, the #2671 fix), but any other fp8/fp4 value falls through to `os << '(' << FlexibleHexFormat(op->value) << 'f' ...` unconditionally ([L5789-L5791](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/src/cuda/codegen/codegen_cuda.cc#L5789)); for a non-finite value that spelling yields `inf`/`nan`, and the trailing `'f'` produces the undefined identifier `inff`/`nanf`. The sibling [fp16/bf16 branch](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/src/cuda/codegen/codegen_cuda.cc#L5730-L5752) and the fp32/fp64 branch both guard `std::isinf`/`std::isnan` first.

The infinity reaches that branch from the reduce lowering: [`MakeInitValue`](https://github.com/tile-ai/tilelang/blob/8001cc4ccf6149382d2019654a19f59c1d4d0482/src/backend/common/op/reduce.h#L207-L231) returns `make_const(dst->dtype, -INFINITY)` for max (L222) and `+INFINITY` for min (L230), in the fp8 destination's dtype. Note `float8_e4m3fn` has no infinity encoding, so for e4m3 a valid init also needs a finite sentinel; e5m2 does have inf and #2671's `::bitcast` approach applies directly.

<details><summary>Reproduced this session (0.1.13, L40S) — got vs control</summary>

```
# dt="float8_e4m3", input a.to(torch.float8_e4m3fn):
COMPILE/RUN EXCEPTION:
   tvm_kernels.cu(23): error: identifier "inff" is undefined
   emitted:  Cf[0] = fp8_e4_t(-inff/*-inf*/);
   ORACLE: FAIL

# dt="float4_e2m1fn" (the accepted float4 spelling), same fp8 inputs:
COMPILE/RUN EXCEPTION:
   tvm_kernels.cu(23): error: identifier "inff" is undefined
   emitted:  tl_fp4_packed_store(Cf_packed, 0, fp4_e2_t(-inff/*-inf*/));
   ORACLE: FAIL

# dt="float32" (control), same fp8 inputs:
maxerr 0.0
   ORACLE: PASS

# dt="float8_e5m2", input a.to(torch.float8_e5m2): the inff crash is FIXED here
#   (#2671 guard applies) — PrintConst now emits  Cf[0] = fp8_e5_t::bitcast(0xfc);
#   but the e5m2 reduce still fails to compile with a DISTINCT error
#   (no operator "=" for `Cf[0] = max(...)`), not the inff path.
# So on 0.1.13 the inff codegen crash is float8_e4m3 and float4_e2m1fn only.
```
</details>

**Suggested fix.** Extend the `std::isinf`/`std::isnan` guard that the fp16/bf16 and fp32/fp64 branches already have to the fp8/fp4 branch of `PrintConst` (`codegen_cuda.cc`). #2671 does this for `float8_e5m2` via `::bitcast` to the inf/nan bit pattern; the remaining `float8_e4m3`/`float4` fall-through still needs it. Because e4m3 has no inf encoding, the reduce init/emitted constant would need to be a representable finite sentinel rather than infinity, so this is more than copying the e5m2 branch verbatim.

**Provenance.** The unguarded fp8/fp4 branch of `PrintConst` was introduced with `[Enhancement] Enhance FP8/FP4 type handling in CUDA codegen` ([#323](https://github.com/tile-ai/tilelang/pull/323), 2025-04-05) and has never guarded inf/nan; the current line positions come from the cache-refactor [#780](https://github.com/tile-ai/tilelang/pull/780) (2025-09-02), which relocated but did not change the logic. Present since #323; not a regression.

**Dedup.** Closely related to [#2670](https://github.com/tile-ai/tilelang/issues/2670) (`T.infinity("float8_e5m2")` internal fatal abort), fixed by [#2671](https://github.com/tile-ai/tilelang/pull/2671). That report/fix is about `T.infinity()`, whose frontend `infinity_op` aborts before codegen, and #2671's codegen change only covers `float8_e5m2`. This report is the codegen `PrintConst` path reached by an ordinary `reduce_max`/`reduce_min` init on `float8_e4m3`/`float4`, which #2671 does not fix and which still crashes on current `main`. I found no other open or closed report of the `float8_e4m3`/`float4` codegen crash.

**Reach.** The trigger is `reduce_max`/`reduce_min` (a common reduction idiom) with the reduce *destination fragment* declared as an fp8 dtype. fp8 reductions appear in the shipped fp8 examples (`examples/deepseek_mla/experimental/example_mla_decode_kv_fp8.py:69` does `T.reduce_max`), but there the destination `scores_max` is the fp32 `accum_dtype` (`:43`), so it takes the guarded fp32 branch and dodges the crash — that is why CI is green. **Verified this session (0.1.13):** built that example's kernel verbatim (default `dtype=float16`, `q_dtype=float8_e4m3fn`, `accum_dtype=float32`); it **compiles cleanly** and its infinity init prints as `-CUDART_INF_F` (guarded fp32 branch) — `inff` does not appear anywhere in the emitted source. The example does *not* exercise the bug precisely because the reduce destination is fp32. The bug fires only when the reduce output fragment is itself `float8_e4m3`/`float4` (or, on 0.1.9, `float8_e5m2`); no shipped example or test declares an fp8 reduce destination.

**Impact.** The narrow trigger is a `reduce_max`/`reduce_min` whose destination fragment is declared as an fp8/fp4 dtype (so the infinity init hits the unguarded `PrintConst` sub-branch). When it fires the failure is a loud, deterministic nvcc compile error (`identifier "inff" is undefined`) caught at build time — it blocks that one kernel from compiling but corrupts no value and cannot slip into a running workload silently. Fixing it closes the fp8/fp4 non-finite-constant printing class so an fp8-typed reduction compiles like the fp16/fp32 destinations already do.

<details><summary>§17 Generalization record</summary>

**Two-level root.**
- **SOURCE-level:** the fp8/fp4 sub-branch of `PrintConst` in `src/cuda/codegen/codegen_cuda.cc` (v0.1.13 L5776–L5792) lacks the `std::isinf`/`std::isnan` special-case that every sibling branch (fp16/bf16 at L5730, tfloat32, fp32/fp64) has. After #2671 added an inf/nan `::bitcast` guard *only* for `float8_e5m2`, any *other* fp8/fp4 non-finite value falls through to `os << '(' << FlexibleHexFormat(op->value) << 'f' ...` (L5789), which spells a non-finite value as `inf`/`nan` + trailing `f` = the undefined C++ identifier `inff`/`nanf`.
- **OPERATOR-level:** any op whose lowering plants a non-finite constant *in an fp8/fp4 dtype* reaches this branch. The only in-tree producer is `MakeInitValue` in `src/backend/common/op/reduce.h` (L207–L231): `reduce_max` → `make_const(dst, -INFINITY)` (L222), `reduce_min` → `+INFINITY` (L230). This `make_const` bypasses the frontend `T.infinity` guard, so reduce is (near-)uniquely the path that gets a non-finite fp8 constant to codegen.

**Found-first:** `reduce_max` into a `float8_e4m3` destination fragment → nvcc `identifier "inff" is undefined`. Class: **missing inf/nan guard in a `PrintConst` dtype sub-branch** (codegen missing-case family).

**4-axis sweep** (fresh process each, 0.1.13 `/tmp/tl113_venv`, L40S sm_89):

| axis | cell tested | observed result | same-root? |
|---|---|---|---|
| found-first | `reduce_max`, e4m3 dst | `Cf[0] = fp8_e4_t(-inff/*-inf*/);` → `identifier "inff" is undefined` | — (the bug) |
| related-operator | `reduce_min`, e4m3 dst (+INF init) | `Cf[0] = fp8_e4_t(inff/*inf*/);` → `identifier "inff" is undefined` | **SAME-ROOT** |
| related-operator | `reduce_sum`, e4m3 dst (zero init) | compile fails but **NO `inff`** (fails on unrelated fp8 `shfl`/`operator<` infra); zero init dodges the non-finite branch | distinct (discriminator: root is the *infinity init*, not fp8-reduce generally) |
| related-type | `float4_e2m1fn`, `reduce_max` dst | `tl_fp4_packed_store(..., fp4_e2_t(-inff ))` → `identifier "inff" is undefined` | **SAME-ROOT** (L5789 fall-through covers `is_float8() \|\| is_float4()`) |
| related-type | `float8_e5m2`, `reduce_max` dst | `Cf[0] = fp8_e5_t::bitcast(0xfc);` (no `inff`), then a **distinct** `no operator "=" matches these operands` | distinct sink (guarded by #2671) |
| related-type | `float32`, `reduce_max` dst (control) | compiles, `maxerr 0.0` | distinct (guarded fp32 branch) |
| similar-logic / related-source | `T.fill(Cf, -T.infinity("float8_e4m3"))` (direct non-finite fp8 const, no reduce) | frontend **fatal**: `Cannot decide infinity for type float8_e4m3` — aborts in `infinity_op` *before* codegen | distinct: this is the #2670 frontend face; it does **not** reach `PrintConst`, confirming reduce's internal `make_const` is the entry point that bypasses the frontend guard |
| related-source (control) | `T.fill(Cf, Cast(e4m3, 1.5))` (finite fp8 const via same fall-through) | compiles cleanly, no error | confirms the L5789 fall-through is correct for *finite* values — only non-finite spelling is broken |

**Example run (Reach).** `examples/deepseek_mla/experimental/example_mla_decode_kv_fp8.py` (the only shipped fp8-reduce example) built verbatim on 0.1.13 → **compiles clean, no `inff`**; its infinity init prints `-CUDART_INF_F` because `scores_max` is fp32 `accum_dtype`. The shipped example genuinely dodges the bug (fp32 reduce destination). No shipped example/test uses an fp8 reduce destination.

**Reframe decision:** kept specific to `reduce_max`/`reduce_min` with fp8/fp4 destination. The class is "non-finite fp8/fp4 constant reaching `PrintConst`", but reduce is the *only* reachable producer (the direct `T.infinity` path is walled off by the frontend #2670 abort), so the trigger boundary is genuinely narrow. e4m3 and float4_e2m1fn are the same bug (one fix). e5m2 (distinct, guarded) and the reduce_sum/fp8-`shfl` infra failure are **distinct, not filed**.

**Distinguished from #2670/#2671:** those are the `T.infinity()` frontend `infinity_op` abort plus the e5m2 codegen `::bitcast` guard; this is the e4m3/float4 codegen `PrintConst` fall-through that #2671 left unguarded.

</details>
